packages/platforms/mellanox/x86-64/x86-64-mlnx-msn2740/onlp/builds/src/module/src/fani.c

```diff
@@ -0,0 +1,450 @@
+    info = &linfo[local_id];
+
+    if (0 == (ONLP_FAN_CAPS_SET_RPM & info->caps)) {
+        return ONLP_STATUS_E_UNSUPPORTED;
+    }
+
+    /* reject rpm=0% (rpm=0%, stop fan) */
+    if (0 == rpm) {
+        return ONLP_STATUS_E_INVALID;
+    }
+
+    /* Set fan speed
+       Converting percent to driver value.
+       Driver accept value in range between 153 and 255.
+       Value 153 is minimum rpm.
+       Value 255 is maximum rpm.
+    */
+    if (local_id > sizeof(min_fan_speed)/sizeof(min_fan_speed[0])) {
+        return ONLP_STATUS_E_INTERNAL;
+    }
+    if (max_fan_speed[local_id] - min_fan_speed[local_id] < 0) {
+        return ONLP_STATUS_E_INTERNAL;
+    }
+    if (rpm < min_fan_speed[local_id] || rpm > max_fan_speed[local_id]) {
+        return ONLP_STATUS_E_PARAM;
+    }
+
+    temp = (rpm - min_fan_speed[local_id]) * (RPM_MAGIC_MAX - RPM_MAGIC_MIN) /
+        (max_fan_speed[local_id] - min_fan_speed[local_id]) + RPM_MAGIC_MIN;
+
+    snprintf(r_data, sizeof(r_data), "%d", (int)temp);
+    nbytes = strnlen(r_data, sizeof(r_data));
+    rv = onlp_file_write((uint8_t*)r_data, nbytes, "%s%s", PREFIX_PATH,
+            fan_path[local_id].r_speed_set);
+	if (rv < 0) {
+		return ONLP_STATUS_E_INTERNAL;
+	}
+
+    return ONLP_STATUS_OK;
+}
+
+/*
+ * This function sets the fan speed of the given OID as a percentage.
+ *
+ * This will only be called if the OID has the PERCENTAGE_SET
+ * capability.
+ *
+ * It is optional if you have no fans at all with this feature.
+ */
+int
+onlp_fani_percentage_set(onlp_oid_t id, int p)
+{
+    float temp = 0.0;
+    int   rv = 0, local_id = 0, nbytes = 10;
+    char  r_data[10]   = {0};
+    onlp_fan_info_t* info = NULL;
+
+    VALIDATE(id);
+    local_id = ONLP_OID_ID_GET(id);
+    info = &linfo[local_id];
+
+    if (0 == (ONLP_FAN_CAPS_SET_PERCENTAGE & info->caps)) {
+        return ONLP_STATUS_E_UNSUPPORTED;
+    }
+
+    /* reject p=0% (p=0%, stop fan) */
+    if (0 == p) {
+        return ONLP_STATUS_E_INVALID;
+    }
+
+    if (p < PERCENTAGE_MIN || p > PERCENTAGE_MAX) {
+        return ONLP_STATUS_E_PARAM;
+    }
+
+    /* Set fan speed
+       Converting percent to driver value.
+       Driver accept value in range between 153 and 255.
+       Value 153 is 60%.
+       Value 255 is 100%.
+    */
+    temp = (p - PERCENTAGE_MIN) * (RPM_MAGIC_MAX - RPM_MAGIC_MIN) /
+        (PERCENTAGE_MAX - PERCENTAGE_MIN) + RPM_MAGIC_MIN;
+
+    snprintf(r_data, sizeof(r_data), "%d", (int)temp);
+    nbytes = strnlen(r_data, sizeof(r_data));
+    rv = onlp_file_write((uint8_t*)r_data, nbytes, "%s%s", PREFIX_PATH,
+            fan_path[local_id].r_speed_set);
+	if (rv < 0) {
+		return ONLP_STATUS_E_INTERNAL;
+	}
+
+    return ONLP_STATUS_OK;
+}
+
+/*
+ * This function sets the fan speed of the given OID as per
+ * the predefined ONLP fan speed modes: off, slow, normal, fast, max.
+ *
+ * Interpretation of these modes is up to the platform.
+ *
+ */
+int
+onlp_fani_mode_set(onlp_oid_t id, onlp_fan_mode_t mode)
```

Comment:
Please remove these unsupported functions. The stubs are not needed.

Reply:
pulling back request and issuing new one with comments.